### ISSUE TITLE
BREAKING: Autoform materialize is not for Meteor version > 1.3.1; if mozfet-materialize >2.0.4

### ISSUE THREAD

**Issue by kwuite** [mozfet/meteor-autoform-materialize#18]
This package is only compatible with Meteor 1.2 - 1.3.1 starting from versions mozfet materialize version > 2.0.4 because the fourseven verison used by [mozfet materialize time-picker](https://github.com/mozfet/meteor-materialize-time-picker/blob/master/package.js), can break at versions above 1.3.1. Especially when you use fourseven in your project because you will have Memory leaks and this is a huge risk for hot code pushing.

We found out our problem after testing the hot code push!

# [Fourseven Compatibility with Meteor](https://github.com/fourseven/meteor-scss)

Meteor Version | Recommended fourseven:scss version
-- | --
1.0 - 1.1 | 3.2.0
1.2 - 1.3.1 | 3.4.2
1.3.2+ | 3.8.0_1
1.4.0 | 3.8.1
1.4.1+ | 4.5.4
1.6+ | 4.5.4

Currently, I am working with Meteor 1.5 and since I had not a strict version restriction on fourseven@4.5.4 this package broke our deployment because it downgraded the fourseven package to version 3.13.0.

```   Linking                                   -
<--- Last few GCs --->

  905828 ms: Mark-sweep 689.8 (735.3) -> 689.8 (735.3) MB, 623.0 / 0 ms [allocation failure] [scavenge might not succeed].
  906444 ms: Mark-sweep 689.8 (735.3) -> 689.8 (735.3) MB, 616.0 / 0 ms [allocation failure] [scavenge might not succeed].
  907066 ms: Mark-sweep 689.8 (735.3) -> 688.6 (735.3) MB, 621.8 / 0 ms [last resort gc].
  907680 ms: Mark-sweep 688.6 (735.3) -> 689.8 (735.3) MB, 614.1 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 3B16CC95 <JS Object>
    1: JSONSerialize(aka JSONSerialize) [native json.js:~120] [pc=5BAA4533] (this=3B1080DD <undefined>,G=3B171681 <String[4]: data>,j=314A07A5 <an Object with m
ap 06ED957D>,v=314A07FD <JS Function replacer (SharedFunctionInfo 3DEF3001)>,w=314A087D <JS Array[2]>,x=3B108159 <String[0]: >,y=3B108159 <String[0]: >)
    2: SerializeObject(aka SerializeObject) [native json.js:97] [pc=17F95214] (this=3B1080...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
```

My Search for Answers:

https://github.com/meteor/meteor/issues/8157
https://github.com/meteor/meteor/issues/7526
https://github.com/Dogfalo/materialize/issues/3446

**Comment by mozfet** [comment]
Thank you for providing the detailed information, sorry to hear this package caused you production issues. Unfortunately I have not yet had the time to test this package with Meteor 1.5 yet... 

I did have a lot of challenges to get fourseven working when I took over this package, and had to restrain the version at times in the past, however had no issues myself using 1.4.4.3 so it seemed to be fixed. Seems I was wrong.

The good news is that Meteor 1.5 made the top of my list for the coming week, and your having a prob problem makes it my first thing for the week.

To solve this issue, tomorrow first thing I will:
 - Update the fourseven version in the timepicker package and test using playground
 - Upgrade the entire suite to Meteor 1.5 and test using playground

If I do not hit any obstacles, there should be a new build by end of tomorrow (CET).

**Comment by kwuite** [comment]
Thanks for the reply.

In order to get my server up and running, I locally imported the time picker and set the fourseven versoin to 4.5.4. Unfortatently we do not use the time picker so I cannot confirm whether it works as it should.

Do keep in mind that you might need a build chain like node-gyp for fourseven to be imported. Also taking a look at the [materializecss ](https://github.com/Dogfalo/materialize) library might be worth your while since they migrated from fourseven to node-sass with grunt I believe.

Well, with confidence in mind I await your reply with a new version;).

**Comment by mozfet** [comment]
Getting rid of fourseven:scss is one thing, but a lot of other users still use poetic:materialize-scss, which uses fourseven:scss... thats why I chose fourseven for the timepicker...

poetic:materialize-scss 1.97.6_1 has dependancy on fourseven:scss 3.4.3... I've not been able to get any later versions of poetic:materialize-scss to play nicely...

looks like this is going to be a bigger job than I thought...

**Comment by mozfet** [comment]
fourseven uses node-sass to compile Sass as well...

**Comment by mozfet** [comment]
poetic:materialize-scss did however not update since April last year, looks stale...

**Comment by mozfet** [comment]
I'm forking poetic:materialize-scss to see if I can upgrade it to Materialize v0.99.0 and foursever:scss; this version of Materialize also includes a time picker, so I can use that instead of maintaining one just for this package...

**Comment by jfhenon** [comment]
Hello, what is the advantage of poetic knowing that materializecss is the official npm package?
it sound to me that a better long term solution is to focus on materialize.  my two cents.

**Comment by kwuite** [comment]
Agreed with @jfhenon on the poetic part. Seems trivial to work on that abandoned package with all the cards on the table. Besides, why upgrade poetic if they can use an older release of your package? 

The materialize package does not conflict with my fourseven even though they use node-sass as well. Going the same route as them would be the smartest way if you ask me.

**Comment by mozfet** [comment]
Could not get poetic:materialise upgraded ... loads of failed tests... bah...

The meteor branch of the materialize package IS poetic:materialize... so it is part of the official package.

The benefits of poetic is that you can modify and overwrite the sass of materialize inside your own meteor project, when you just use the materialize package you get only compiled CSS in the package and cannot modify or overwrite the SCSS...

**Comment by mozfet** [comment]
Using an older version of this package will not help others trying to migrate to Meteor 1.5, since the dependency chain just gets older and more conflicting with newer packages.

**Comment by mozfet** [comment]
In order to move forward: I am going to update the timepicker to make use of the new Materialize 0.99.0 timepicker. This will remove this suite's dependancy on fourseven completely.

Because poetic only supports up to materialize 0.96 at the moment, the timepicker will not work for those using poetic:materialize-scss; sorry its the best I can do at this stage...

Those that use poetic:materialize-scss (including myself) will need to wait until I can get it properly upgraded, or user older versions of this suite until that is sorted out to work properly with the latest version of fourseven.

**Comment by kwuite** [comment]
> The benefits of poetic is that you can modify and overwrite the sass of materialize inside your own meteor project, when you just use the materialize package you get only compiled CSS in the package and cannot modify or overwrite the SCSS...

The above is actually incorrect. In my project I am using the material library since V0.97. With the npm package you get the compiled and uncompiled distribution of the sass files. 

```
client
------theme
-----------_materialize.scss
------main.scss

_materialize.scss =>

// Mixins
@import "{}/node_modules/materialize-css/sass/components/mixins";
@import "{}/node_modules/materialize-css/sass/components/color";

// Variables
@import "variables";
@import "{}/node_modules/materialize-css/sass/components/variables";

// Reset
@import "{}/node_modules/materialize-css/sass/components/normalize";

// components
@import "{}/node_modules/materialize-css/sass/components/global";
@import "{}/node_modules/materialize-css/sass/components/icons-material-design";
@import "{}/node_modules/materialize-css/sass/components/badges";

main.scss =>

// Import Materialize CSS / Components
@import "theme/materialize";

```

The above strategy works for the javascript files as well, so In our project, we can use `@extend and @include` however we like, we pick the sass and js files and our Fourseven version and minification libraries will do the project bundling.

**Clockpicker**

Why reinvent the wheel right? We have been using the[ `"materialize-clockpicker": "^1.0.1"`](https://github.com/chingyawhao/materialize-clockpicker) which is an extension for the materialize library. I honestly didn't know you had a library until I saw you imported it in your main package and I traced back our Fourseven conflict to your time-picker package.

**Comment by mozfet** [comment]
I just published V2.1.0 on Atmosphere.

I've tested it with Materialize 0.99.0 on Meteor 1.4, and changed pickatime to use the newly introduced Materialize timepicker. I did update the playground but did not do a full regression test though, please let me know if you come across new issues.

This should solve any issues due to fourseven:scss dependancies, and associated hot code push issues.

**Comment by mozfet** [comment]
```
$ for p in `meteor list | grep '^[a-z]' | awk '{ print $1"@"$2 }'`; do echo "$p"; meteor show "$p" | grep -E '^  [a-z]'; echo; done
aldeed:autoform@6.2.0
  aldeed:collection2-core@2.0.0 (weak dependency)
  aldeed:moment-timezone@0.4.0 (weak dependency)
  blaze@2.1.5
  deps@1.0.10
  ecmascript@0.4.1
  ejson@1.0.9
  isobuild:isopack-2@1.0.0
  jquery@1.11.6
  livedata@1.0.16
  momentjs:moment@2.10.6
  mrt:moment-timezone@0.2.1 (weak dependency)
  random@1.0.7
  reactive-dict@1.1.5
  reactive-var@1.0.7
  reload@1.1.6 (weak dependency)
  templating@1.1.7
  ui@1.0.9
  underscore@1.0.6

aldeed:collection2-core@2.0.1
  ecmascript@0.6.1
  ejson@1.0.0
  insecure@1.0.0 (weak dependency)
  isobuild:isopack-2@1.0.0
  minimongo@1.0.0
  mongo@1.0.4
  raix:eventemitter@0.1.3
  tmeasday:check-npm-versions@0.3.1
  underscore@1.0.0

autopublish@1.0.7

blaze-html-templates@1.1.2
  blaze@2.3.2
  spacebars@1.0.15
  templating@1.3.2
  ui@1.0.12

bundle-visualizer@1.0.3
  dynamic-import@0.1.1
  ecmascript@0.8.0
  isobuild:dynamic-import@1.5.0
  isobuild:isopack-2@1.0.0

crystalhelix:wnumb@1.0.1
  ecmascript@0.1.6
  isobuild:isopack-2@1.0.0

dynamic-import@0.1.1
  browser-policy-content@1.1.0 (weak dependency)
  check@1.2.5
  ddp@1.2.5
  ecmascript@0.8.0
  isobuild:dynamic-import@1.5.0
  isobuild:isopack-2@1.0.0
  modules@0.9.1
  promise@0.8.9

ecmascript@0.8.1
  babel-compiler@6.19.3
  babel-runtime@1.0.1
  ecmascript-runtime@0.4.1
  isobuild:compiler-plugin@1.0.0
  isobuild:isopack-2@1.0.0
  modules@0.9.2
  promise@0.8.9

es5-shim@4.6.15
  modules@0.7.7

insecure@1.0.7

jquery@1.11.10
  modules@0.7.7

materialize:materialize@0.99.0
  fourseven:scss@3.10.0
  isobuild:isopack-2@1.0.0
  jquery@1.0.1

meteor-base@1.1.0
  ddp@1.2.5
  dynamic-import@0.1.0
  hot-code-push@1.0.4
  livedata@1.0.18
  underscore@1.0.10
  webapp@1.3.16

mobile-experience@1.0.4
  fastclick@1.0.11
  launch-screen@1.0.11
  mobile-status-bar@1.0.12

mongo@1.1.18
  allow-deny@1.0.5
  autopublish@1.0.7 (weak dependency)
  binary-heap@1.0.10
  callback-hook@1.0.10
  check@1.2.5
  ddp@1.2.5
  diff-sequence@1.0.7
  disable-oplog@1.0.7 (weak dependency)
  ecmascript@0.8.0
  ejson@1.0.13
  facts@1.0.9 (weak dependency)
  insecure@1.0.7 (weak dependency)
  isobuild:isopack-2@1.0.0
  minimongo@1.2.0
  mongo-id@1.0.6
  npm-mongo@2.2.24
  random@1.0.10
  tracker@1.1.3
  underscore@1.0.10
  webapp@1.3.16 (weak dependency)
```

**Comment by mozfet** [comment]
LOL: Materialize 0.99.0 uses fourseven:scss@3.10.0!!!!

I see this in the package tree after clean install of playground...

**Comment by mozfet** [comment]
@kwuite I did not know that was possible (using scss directly from materialize package)! I will try it out immediately. This is really embarrassing... I though I was becoming a MaterializeCSS ninja, and you put me in my place... you should help out here if you have time...

**Comment by mozfet** [comment]
@kwuite I only now notice you are using the npm materialize package, instead of the atmosphere package, I have not even though of doing that yet, but now that you mention it, it seems the right way to go. I will change the playground to do it this way as well (and then slowly the rest of my projects)...

May I ask what strategy do you use to load the MaterialiseCSS JS?

**Comment by mozfet** [comment]
I've been trying for 3 hours to get Materialize working in Meteor using NPM package. The web has very little on how to do this... most people run when mentioning importing scss into meteor from a npm repo...

So lets assume I want to add everything in materialize into a new project: I tried the following:

Your scss import statements did not work for me, Meteor could not see the scss files. So I created an scss-config.json with:
```
{
  "includePaths": [
    "{}/node_modules/materialize-css/sass/"
  ]
}
``` 
- in main.scss: ```@import "materialize.scss"``` which at least allowed fourseven to see the scss (I do not know why...)
-in startup/client/startup/index.js `import './materialize.js'`
-in startup/client/startup/materialize.js:
- `$ meteor npm install hammerjs --save', cause materialize does not depend on it...
```
import 'materialize-css/dist/js/materialize.js';
import 'materialize-css/extras/noUISlider/nouislider.js';
//the nouisliders's css should probably be imported here?
```

now I got an error:
```
While processing files with fourseven:scss (for target web.browser):
   /client/main.scss: Scss compiler error: File to import: input-fields not found in file: components/forms/forms
```
before this I had a hell of time trying to figure out the sass import sequences... wasted time there as well... I'm about to give up.. and I have not even started looking at adding the fonts and the icons... its all very m.a.n.u.a.l....

** To me, this is precisely why poetic:scss was so nice. You start a new meteor project, add fourseven:scss and poetic:materialize-scss, and boom the entirety of materialize is installed and you can use it... pardon the pun, its poetic... Or do I have the whole thing wrong?**

**Comment by kwuite** [comment]
My sincere apologies for replying incredibly later, I am for a project in another country and I do not have allot of time to focus on anything else then "running the project". I'll be back next week Monday, then I can write a demo app with the way we use meteor + materialize + imports if it's not too late for you.

**Comment by mozfet** [comment]
@kwuite A demo app would be fantastic! If you ever get some free time going forward, I really can do with some help here :)

I'm going to close this issue; please open a new issue and label it as improvement with the directions for your demo. Thank you very much.

**Comment by mozfet** [comment]
Hi Kelvin,

Did you get some time to look into this?

Kind Regards
Wynand

**Comment by kwuite** [comment]
Dear Wynand,

Currently, I am on holiday so I'm "trying" not to commit myself to anything else than my family. I'll get back the 8/9th of August to the Netherlands. Once I'm back I will download a random Meteor boilerplate (if suggestions then please share) and implement the strategy that we use internally for materialize as discussed earlier into this boilerplate and I will share the code with you.

Would that be a way that I can help you?

Cine Cera,

Kelvin

**Comment by mozfet** [comment]
Oh, sorry to intrude on your holiday, of course, focus on your family! Any boilerplate will do, I just need an example. Its not critical right now, no one is screaming for it :)

**Comment by mozfet** [comment]
@kwuite have you had some time to look at this? I do not get further than Materialize.js not finding Hammer... and no-one on the net seems to have solved that... and it seems I really need to get rid of poetic:materializescss in all my projects... I need to have sass in order to customise the materialise primary and secondary colors...

**Comment by jfhenon** [comment]
Sorry I didn't participate more but we are moving now to another UI framework (Semantic). 
To respond to your question, did you add the following client code to make sure you make Hammer a global variable as well as Materialize? 
`
import _Hammer from 'hammerjs';      
Hammer = _Hammer;      
Materialize = window.Materialize;      
`

**Comment by mozfet** [comment]
Got it working by just importing hammer… hammer is included in the materialize distro, but I could not get that working so I installed hammer using NPM. I also had to move to fonds from the materialize package and dropped it into public. It seems to work, but I’ll do a full regression test today...

**Comment by mozfet** [comment]
Finally this item can be closed. Fixed in 'v3.0.1'. This version removes support for  poetic:materialize-scss and has been tested in the playground using NPM installation of Materialize-css and using fourseven:scss and SASS to override primary and secondary colours. See the updated readme for instructions on installing materialize on meteor using NPM.
